### ISSUE TITLE
[install] npm7 and above installation project dependencies on error warning to use the legacy-peer-deps parameter

### ISSUE THREAD

**Issue by me20150505** [npm/cli#5925]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

Run the npm install command directly to install the dependencies, and the error "could not resolve" is reported to terminate the dependencies installation.

Compared with using yarn install command to install the dependencies, the project dependencies can be installed normally, and there is no situation that the dependencies installation cannot be installed.

Moreover, because the preinstall hook does not work properly ([preinstall not works](https://github.com/npm/cli/issues/5892)), it is impossible to set the command to configure the legacy-peer-deps parameter in preinstall hook.

### Expected Behavior

Is there any way that npm install command can install dependencies normally?

### Steps To Reproduce

1. In this environment...
2. With this config...
3. Run '...'
4. See error...

### Environment

- npm: 9.1.3
- Node.js: v18.12.1
- OS Name: Ubuntu 20.04.4 LTS
- System Model Name: Linux version 5.4.0
- npm config:
```ini
; "user" config from /home/yfw/.npmrc

registry = "https://repo.huaweicloud.com/repository/npm/"
strict-ssl = false

; "cli" config from command line options

location = "project"

; node bin location = /home/yfw/bin/node-v18.12.1-linux-x64/bin/node
; node version = v18.12.1
; npm local prefix = /home/yfw/code/compiler
; npm version = 9.1.3
; cwd = /home/yfw/code/compiler
; HOME = /home/yfw
; Run `npm config ls -l` to show all defaults.
```

**Comment by nlf** [comment]
a "could not resolve" error means that your package tree is invalid in some way. using `--legacy-peer-deps` can sometimes work around this problem, but it is not a cure all and can lead to unexpected behaviors. if you can share the full error message that you're receiving, we can likely help you determine another way to resolve this problem. `yarn` being able to install this dependency tree is because `yarn` lacks proper support for peer dependencies.

> Moreover, because the preinstall hook does not work properly (https://github.com/npm/cli/issues/5892), it is impossible to set the command to configure the legacy-peer-deps parameter in preinstall hook.

even if the preinstall hook worked as you want it to, it would not work for setting a parameter for the current npm installation. we only read configs before we start installing things, and once they're in progress even if you modified the user's `.npmrc` to set the flag, it would not take effect for the current install.

we need more information about what specifically is breaking for you. once we have that, we can help you determine a fix.

**Comment by me20150505** [comment]
Thanks for your reply.

When I use npm install command to install dependencies, the error is as follows:

```bash
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR!
npm ERR! While resolving: ts-node@9.1.1
npm ERR! Found: typescript@4.2.3-r2
npm ERR! node_modules/typescript
npm ERR!   typescript@"file:./deps/ohos-typescript-4.2.3-r2.tgz" from the root project
npm ERR!   peer typescript@"*" from ts-loader@8.4.0
npm ERR!   node_modules/ts-loader
npm ERR!     ts-loader@"^8.0.12" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer typescript@">=2.7" from ts-node@9.1.1
npm ERR! node_modules/ts-node
npm ERR!   dev ts-node@"^9.1.1" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: typescript@4.9.3
npm ERR! node_modules/typescript
npm ERR!   peer typescript@">=2.7" from ts-node@9.1.1
npm ERR!   node_modules/ts-node
npm ERR!     dev ts-node@"^9.1.1" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR!
npm ERR! For a full report see:
npm ERR! /home/yfw/.npm/_logs/2022-12-07T09_49_47_492Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/yfw/.npm/_logs/2022-12-07T09_49_47_492Z-debug-0.log
```

Is there a way to make the installation of npm install command proceed normally without adding the legacy-peer-deps parameter?

Or can the dependencies installation proceed normally only by adding the legacy-peer-deps parameter (that to use `npm install --legacy-peer-deps`)?

**Comment by ljharb** [comment]
@me20150505 yes - you can make your dep graph valid. It's currently invalid, because a prerelease of typescript can't satisfy any non-prerelease ranges. If you use a non-prerelease version of TypeScript, it will work.

**Comment by me20150505** [comment]
A complete log of this run can be found as follows:

```log
0 verbose cli /home/yfw/bin/node-v18.12.1-linux-x64/bin/node /home/yfw/bin/node-v18.12.1-linux-x64/bin/npm
1 info using npm@9.2.0
2 info using node@v18.12.1
3 timing npm:load:whichnode Completed in 1ms
4 timing config:load:defaults Completed in 2ms
5 timing config:load:file:/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/npmrc Completed in 1ms
6 timing config:load:builtin Completed in 1ms
7 timing config:load:cli Completed in 2ms
8 timing config:load:env Completed in 1ms
9 timing config:load:file:/home/yfw/code/compiler/.npmrc Completed in 0ms
10 timing config:load:project Completed in 1ms
11 timing config:load:file:/home/yfw/.npmrc Completed in 3ms
12 timing config:load:user Completed in 3ms
13 timing config:load:file:/home/yfw/bin/node-v18.12.1-linux-x64/etc/npmrc Completed in 0ms
14 timing config:load:global Completed in 0ms
15 timing config:load:setEnvs Completed in 1ms
16 timing config:load Completed in 11ms
17 timing npm:load:configload Completed in 11ms
18 timing npm:load:mkdirpcache Completed in 1ms
19 timing npm:load:mkdirplogs Completed in 0ms
20 verbose title npm install
21 verbose argv "install"
22 timing npm:load:setTitle Completed in 2ms
23 timing config:load:flatten Completed in 4ms
24 timing npm:load:display Completed in 5ms
25 verbose logfile logs-max:10 dir:/home/yfw/.npm/_logs/2022-12-08T04_09_10_162Z-
26 verbose logfile /home/yfw/.npm/_logs/2022-12-08T04_09_10_162Z-debug-0.log
27 timing npm:load:logFile Completed in 6ms
28 timing npm:load:timers Completed in 0ms
29 timing npm:load:configScope Completed in 0ms
30 timing npm:load Completed in 26ms
31 silly logfile done cleaning log files
32 timing config:load:flatten Completed in 1ms
33 timing arborist:ctor Completed in 1ms
34 timing idealTree:init Completed in 515ms
35 timing idealTree:userRequests Completed in 0ms
36 silly idealTree buildDeps
37 timing idealTree:#root Completed in 1ms
38 silly fetch manifest acorn@^8
39 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/acorn 172ms (cache miss)
40 silly placeDep ROOT acorn@8.8.1 REPLACE for: acorn-import-assertions@1.8.0 want: ^8
41 timing idealTree:node_modules/acorn-import-assertions Completed in 190ms
42 timing idealTree:node_modules/acorn Completed in 0ms
43 silly fetch manifest acorn@^7.4.0
44 silly placeDep node_modules/espree acorn@7.4.1 OK for: espree@7.3.1 want: ^7.4.0
45 timing idealTree:node_modules/espree Completed in 6ms
46 silly fetch manifest typescript@>=2.7
47 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/typescript 737ms (cache miss)
48 timing idealTree Completed in 1629ms
49 timing command:install Completed in 1710ms
50 verbose stack Error: could not resolve
50 verbose stack     at PlaceDep.failPeerConflict (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/place-dep.js:502:25)
50 verbose stack     at new PlaceDep (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/place-dep.js:156:21)
50 verbose stack     at /home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:942:31
50 verbose stack     at Array.map (<anonymous>)
50 verbose stack     at [buildDepStep] (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:942:8)
50 verbose stack     at async Arborist.buildIdealTree (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:206:7)
50 verbose stack     at async Promise.all (index 1)
50 verbose stack     at async Arborist.reify (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/reify.js:156:5)
50 verbose stack     at async Install.exec (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/lib/commands/install.js:146:5)
50 verbose stack     at async module.exports (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/lib/cli.js:133:5)
51 verbose cwd /home/yfw/code/compiler
52 verbose Linux 5.4.0-126-generic
53 verbose node v18.12.1
54 verbose npm  v9.2.0
55 error code ERESOLVE
56 error ERESOLVE could not resolve
57 error
58 error While resolving: ts-node@9.1.1
58 error Found: typescript@4.2.3-r2
58 error node_modules/typescript
58 error   typescript@"file:./deps/ohos-typescript-4.2.3-r2.tgz" from the root project
58 error   peer typescript@"*" from ts-loader@8.4.0
58 error   node_modules/ts-loader
58 error     ts-loader@"^8.0.12" from the root project
58 error
58 error Could not resolve dependency:
58 error peer typescript@">=2.7" from ts-node@9.1.1
58 error node_modules/ts-node
58 error   dev ts-node@"^9.1.1" from the root project
58 error
58 error Conflicting peer dependency: typescript@4.9.4
58 error node_modules/typescript
58 error   peer typescript@">=2.7" from ts-node@9.1.1
58 error   node_modules/ts-node
58 error     dev ts-node@"^9.1.1" from the root project
58 error
58 error Fix the upstream dependency conflict, or retry
58 error this command with --force or --legacy-peer-deps
58 error to accept an incorrect (and potentially broken) dependency resolution.
59 error
59 error
59 error For a full report see:
59 error /home/yfw/.npm/_logs/2022-12-08T04_09_10_162Z-eresolve-report.txt
60 verbose exit 1
61 timing npm Completed in 1767ms
62 verbose unfinished npm timer reify 1670472550213
63 verbose unfinished npm timer reify:loadTrees 1670472550287
64 verbose unfinished npm timer idealTree:buildDeps 1670472550805
65 verbose unfinished npm timer idealTree:node_modules/ts-node 1670472551003
66 verbose code 1
67 error A complete log of this run can be found in:
67 error     /home/yfw/.npm/_logs/2022-12-08T04_09_10_162Z-debug-0.log
```

**Comment by me20150505** [comment]
> @me20150505 yes - you can make your dep graph valid. It's currently invalid, because a prerelease of typescript can't satisfy any non-prerelease ranges. If you use a non-prerelease version of TypeScript, it will work.

Thank you for your reply.

Typescript may be a customized version at present, I'll try it.

**Comment by me20150505** [comment]
> > @me20150505 yes - you can make your dep graph valid. It's currently invalid, because a prerelease of typescript can't satisfy any non-prerelease ranges. If you use a non-prerelease version of TypeScript, it will work.
> 
> Thank you for your reply.
> 
> Typescript may be a customized version at present, I'll try it.

It still doesn't work properly.

```log
0 verbose cli /home/yfw/bin/node-v18.12.1-linux-x64/bin/node /home/yfw/bin/node-v18.12.1-linux-x64/bin/npm
1 info using npm@9.2.0
2 info using node@v18.12.1
3 timing npm:load:whichnode Completed in 1ms
4 timing config:load:defaults Completed in 3ms
5 timing config:load:file:/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/npmrc Completed in 1ms
6 timing config:load:builtin Completed in 1ms
7 timing config:load:cli Completed in 2ms
8 timing config:load:env Completed in 1ms
9 timing config:load:file:/home/yfw/work/SDK/js/build-tools/ace-loader/.npmrc Completed in 1ms
10 timing config:load:project Completed in 2ms
11 timing config:load:file:/home/yfw/.npmrc Completed in 2ms
12 timing config:load:user Completed in 2ms
13 timing config:load:file:/home/yfw/bin/node-v18.12.1-linux-x64/etc/npmrc Completed in 1ms
14 timing config:load:global Completed in 1ms
15 timing config:load:setEnvs Completed in 1ms
16 timing config:load Completed in 13ms
17 timing npm:load:configload Completed in 13ms
18 timing npm:load:mkdirpcache Completed in 1ms
19 timing npm:load:mkdirplogs Completed in 0ms
20 verbose title npm install
21 verbose argv "install"
22 timing npm:load:setTitle Completed in 2ms
23 timing config:load:flatten Completed in 4ms
24 timing npm:load:display Completed in 6ms
25 verbose logfile logs-max:10 dir:/home/yfw/.npm/_logs/2022-12-09T06_37_58_389Z-
26 verbose logfile /home/yfw/.npm/_logs/2022-12-09T06_37_58_389Z-debug-0.log
27 timing npm:load:logFile Completed in 6ms
28 timing npm:load:timers Completed in 1ms
29 timing npm:load:configScope Completed in 0ms
30 timing npm:load Completed in 30ms
31 silly logfile done cleaning log files
32 timing config:load:flatten Completed in 1ms
33 timing arborist:ctor Completed in 1ms
34 timing arborist:ctor Completed in 0ms
35 timing idealTree:init Completed in 21ms
36 timing idealTree:userRequests Completed in 1ms
37 silly idealTree buildDeps
38 silly fetch manifest @babel/cli@^7.8.4
39 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/@babel%2fcli 366ms (cache miss)
40 silly fetch manifest @babel/core@^7.9.0
41 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/@babel%2fcore 68ms (cache miss)
42 silly fetch manifest @babel/plugin-proposal-class-properties@^7.8.3
43 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/@babel%2fplugin-proposal-class-properties 55ms (cache miss)
44 silly fetch manifest @babel/plugin-transform-runtime@^7.12.10
45 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/npm 927ms (cache miss)
46 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/@babel%2fplugin-transform-runtime 702ms (cache miss)
47 silly fetch manifest @babel/preset-env@^7.9.0
48 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/@babel%2fpreset-env 47ms (cache miss)
49 silly fetch manifest @babel/runtime@^7.0.0
50 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/@babel%2fruntime 61ms (cache miss)
51 silly fetch manifest babel-loader@^8.0.6
52 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/babel-loader 76ms (cache miss)
53 silly fetch manifest webpack@5.48.0
54 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/webpack 336ms (cache miss)
55 silly fetch manifest babel-plugin-require-context-hook@^1.0.0
56 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/babel-plugin-require-context-hook 258ms (cache miss)
57 silly fetch manifest copy-webpack-plugin@^8.1.0
58 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/copy-webpack-plugin 61ms (cache miss)
59 silly fetch manifest css@^3.0.0
60 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/css 322ms (cache miss)
61 silly fetch manifest css-loader@^3.4.2
62 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/css-loader 422ms (cache miss)
63 silly fetch manifest deccjsunit@^1.0.8
64 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/deccjsunit 31ms (cache miss)
65 silly fetch manifest escodegen@^2.0.0
66 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/escodegen 91ms (cache miss)
67 silly fetch manifest esprima@^4.0.1
68 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/esprima 30ms (cache miss)
69 silly fetch manifest hash-sum@^1.0.2
70 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/hash-sum 84ms (cache miss)
71 silly fetch manifest jimp@^0.12.1
72 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/jimp 423ms (cache miss)
73 silly fetch manifest less@^3.11.1
74 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/less 66ms (cache miss)
75 silly fetch manifest less-loader@^5.0.0
76 http fetch GET 200 https://repo.huaweicloud.com/repository/npm/less-loader 351ms (cache miss)
77 silly fetch manifest webpack@^2.0.0 || ^3.0.0 || ^4.0.0
78 timing idealTree Completed in 4239ms
79 timing command:install Completed in 4326ms
80 verbose stack Error: unable to resolve dependency tree
80 verbose stack     at [failPeerConflict] (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:1378:25)
80 verbose stack     at [loadPeerSet] (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:1344:34)
80 verbose stack     at async [buildDepStep] (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:928:11)
80 verbose stack     at async Arborist.buildIdealTree (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:206:7)
80 verbose stack     at async Promise.all (index 1)
80 verbose stack     at async Arborist.reify (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/reify.js:156:5)
80 verbose stack     at async Install.exec (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/lib/commands/install.js:146:5)
80 verbose stack     at async module.exports (/home/yfw/bin/node-v18.12.1-linux-x64/lib/node_modules/npm/lib/cli.js:133:5)
81 verbose cwd /home/yfw/work/SDK/js/build-tools/ace-loader
82 verbose Linux 5.4.0-126-generic
83 verbose node v18.12.1
84 verbose npm  v9.2.0
85 error code ERESOLVE
86 error ERESOLVE unable to resolve dependency tree
87 error
88 error While resolving: ace-loader@1.0.11
88 error Found: webpack@5.48.0
88 error node_modules/webpack
88 error   webpack@"5.48.0" from the root project
88 error
88 error Could not resolve dependency:
88 error peer webpack@"^2.0.0 || ^3.0.0 || ^4.0.0" from less-loader@5.0.0
88 error node_modules/less-loader
88 error   less-loader@"^5.0.0" from the root project
88 error
88 error Fix the upstream dependency conflict, or retry
88 error this command with --force or --legacy-peer-deps
88 error to accept an incorrect (and potentially broken) dependency resolution.
89 error
89 error
89 error For a full report see:
89 error /home/yfw/.npm/_logs/2022-12-09T06_37_58_389Z-eresolve-report.txt
90 verbose exit 1
91 timing npm Completed in 4388ms
92 verbose unfinished npm timer reify 1670567878447
93 verbose unfinished npm timer reify:loadTrees 1670567878527
94 verbose unfinished npm timer idealTree:buildDeps 1670567878550
95 verbose unfinished npm timer idealTree:#root 1670567878551
96 verbose code 1
97 error A complete log of this run can be found in:
97 error     /home/yfw/.npm/_logs/2022-12-09T06_37_58_389Z-debug-0.log
```

**Comment by v-mragunandhan24** [comment]
**Closing** due to age. Adding the comment here for future readers. The less-loader package expects webpack versions 2, 3 or 4. Kindly update the package and see if it fixes the issue. Also please upgrade to the latest npm version. If this is still a problem please feel free to reopen this issue, or create a new issue w/ steps to reproduce.
